### ISSUE TITLE
Developers use IDEs, not UI dashboards. Any way workspace shutdown reminders can be sent as notifications in the IDE?

### ISSUE THREAD

**Issue by mtm20176** [coder/coder#2231]
A v1 customer uses SSH with VS Code local, not code-server, so the devs rarely are in the Coder UI. They will not notice a reminder their workspace is about to shut down.

Devs could just keep their workspaces on indefinitely, but that is cost-prohibitive, especially if a VM.

Are there engineering ideas to accommodate this?

**Comment by ammario** [comment]
I agree that our bump banner in the UI will only catch a small percentage of annoying autostops. Instead of chasing down every IDE and access point, I made a proposal that I think is adequate for this customer and others [here](https://github.com/coder/coder/issues/2233).

**Comment by ketang** [comment]
My current thinking is that we do this as part of a VS Code plugin.

Interim ideas... have we figured out if we can announce to all terminals?  That would include the VS Code terminal if they have it open.  Maybe there are some extra arguments we can add to the ProxyCommand that *hand wave*?

Maybe we can also add something to the CLI like `coder workspace monitor` that is a persistent terminal process that occasionally spits out updates and also warns if shutdown (with terminal bell), basically the CLI equivalent of the popup, and tell customers in our remote VS Code and other IDE instructions to launch that in the terminal.

But also I don't think this is terrible for now.  They can also disable auto off.

**Comment by johnstcn** [comment]
> A v1 customer uses SSH with VS Code local, not code-server, so the devs rarely are in the Coder UI. They will not notice a reminder their workspace is about to shut down.

Note: If using SSH with VS Code Local, they will get a system-native notification sent by the CLI 30 minutes before the scheduled shutdown. However, this does not handle the case when they have the VSCode Web version open and do not have the Coder dashbaord open. In the past, we've already ruled out browser notifications for this purpose as folks tend to find them annoying.

**Comment by ketang** [comment]
Does this announcement show up in the VS Code terminal in remote and web modes?  Telling them to have a terminal open doesn't seem terrible.  Anything other than a  plugin is going to have multiple tradeoffs.

**Comment by johnstcn** [comment]
> Does this announcement show up in the VS Code terminal in remote and web modes? Telling them to have a terminal open doesn't seem terrible. Anything other than a plugin is going to have multiple tradeoffs.

No, it shows up as a system-native desktop notification:
- On Windows, it shows up as a native Toast notification
- On OSX, it shows up as a notification (via `osascript`)
- On Linux, it does whatever your desktop environment does when it gets a `notify-send` via DBus.

However, if you have some sort of focus mode enabled, your desktop environment may end up hiding that notification from you until it's too late.

**Comment by ketang** [comment]
Oooh got it.  I was fixated on terminal announce.  Does that work as we would want?

I think there's a rabbit hole of infinite depth if we get into the "what if they didn't see the notification popup?"  I'm inclined to say that's sufficient.  If it's not, people will tell us.

**Comment by ketang** [comment]
> if we can figure that out.

I'm not disagreeing there, but there are two hard parts.  One is recognizing inactivity, and the other is appropriate delivery of warnings to the user of imminent shutdown.  I think there are more important things to work on before we go further down that path.
